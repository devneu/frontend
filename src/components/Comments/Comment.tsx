import { ReactNode } from 'react'
import { Comment as AntdComment, Avatar } from 'antd';
import { comment } from '../../types/blogTypes';
import scrollToElement from '../../utils/scrollToElement'

interface CommentProps {
   comment: comment,
   children?: ReactNode,
   child?: comment[] | null,
   onSelectReplyTo: (commentId: number | null) => void;
}

const Comment = ({ comment, onSelectReplyTo }: CommentProps) => {
   const { commentText, userName, userAvatar, commentId } = comment;
   const [firstName, lastName] = userName.split(' ');

   const childrenList = comment.children.map((child: comment) => {
      return <Comment
         key={child.commentId}
         comment={child}
         onSelectReplyTo={onSelectReplyTo} />
   })
   return (
      <AntdComment
         actions={[<span
            key="comment-nested-reply-to"
            onClick={() => {
               onSelectReplyTo(commentId)
               scrollToElement('comments__form')
            }}>
            Reply to {firstName}</span>]}
         author={<span className="comments__user-name">{userName}</span>}
         avatar={
            <Avatar className="comments__user-avatar"
               src={userAvatar}
               alt={userName}
            />
         }
         content={
            <p>{commentText}</p>
         }
      >
         {childrenList}
      </AntdComment>
   )
}

export default Comment;