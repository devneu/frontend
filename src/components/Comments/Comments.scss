.comments{
   padding: 15px;
   background-color: #f8f9fa;
   border: 1px solid rgba(0, 0, 0, 0.125);
   border-radius: 5px;
   max-width: 900px;
   margin: 20px auto 0;
   &__list{
      .comments__commentor-name{
         font-weight: 700;
         font-size: 1rem;
         color: #212529;
      }
      .comments__commentor-avatar{
         width: 50px;
         height: 50px;
         border-radius: 50%;
         img{
            width: 100%;
            height: 100%;
         }
      }
   }
}