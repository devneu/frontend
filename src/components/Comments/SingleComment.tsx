import React from 'react';
import { Comment as AntdComment, Avatar } from 'antd';
import { Comment } from '../../types/blogTypes';
import scrollToElement from '../../utils/scrollTo';

interface CommentProps {
  comment: Comment;
  onSelectReplyTo: (commentId: number | null) => void;
}

function SingleComment({ comment: com, onSelectReplyTo }: CommentProps) {
  const { commentText, userName, userAvatar, commentId } = com;
  const childrenList = com.children?.map((child: Comment) => (
    <SingleComment key={child.commentId} comment={child} onSelectReplyTo={onSelectReplyTo} />
  ));
  return (
    <AntdComment
      actions={[
        <button
          type="submit"
          key="comment-nested-reply-to"
          className="comments__reply-btn"
          onClick={() => {
            onSelectReplyTo(commentId);
            scrollToElement('comments__form');
          }}
        >
          Reply to
        </button>,
      ]}
      author={<span className="comments__user-name">{userName}</span>}
      avatar={<Avatar className="comments__user-avatar" src={userAvatar} alt={userName} />}
      content={<p>{commentText}</p>}
    >
      {childrenList}
    </AntdComment>
  );
}

export default SingleComment;
