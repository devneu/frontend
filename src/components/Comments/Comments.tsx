import React, { ReactNode } from 'react';
import { Comment, Avatar, Form, Button, List, Input } from 'antd';
import './Comments.scss'

const { TextArea } = Input;

interface comment {
   postID: number,
   commentID: number,
   replyTo?: number | null,
   replyes?: comment[] | null,
   commentText: string,
   commenterName: string,
   commenterAvatar?: string | null,
}
interface SingleCommentProps {
   comment: comment,
   children?: ReactNode,
}

const SingleComment = ({ children, comment }: SingleCommentProps) => {
   const { commentText, commenterName, commenterAvatar } = comment;
   const commenterFirstName: string = commenterName.split(' ')[0];
   return (
      <Comment
         actions={[<span key="comment-nested-reply-to">Reply to {commenterFirstName}</span>]}
         author={<span className="comments__commentor-name">{commenterName}</span>}
         avatar={
            <Avatar className="comments__commentor-avatar"
               src={commenterAvatar}
               alt={commenterName}
            />
         }
         content={
            <p>{commentText}</p>
         }
      >
         {children}
      </Comment>
   )
}

function showComments(commentsList: any) {
   return commentsList.map((comment: comment) => {
      const { replyes, commentID } = comment
      if (Array.isArray(replyes)) {
         return (
            <SingleComment
               key={commentID}
               comment={comment}>
               {showComments(comment.replyes)}
            </SingleComment>
         )
      }
      return <SingleComment
         key={commentID}
         comment={comment} />
   })
}

const Comments = ({ commentsList }: any) => {
   return (
      <div className="comments">
         <Form className="comments__form">
            <Form.Item>
               <TextArea rows={4} />
            </Form.Item>
            <Form.Item>
               <Button htmlType="submit" type="primary">
                  Add Comment
               </Button>
            </Form.Item>
         </Form>

         <List className="comments__list">
            {
               showComments(commentsList)
            }
         </List>
      </div>
   )
}

export default Comments