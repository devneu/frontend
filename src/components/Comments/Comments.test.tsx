import Comments from './Comments';
import renderer from 'react-test-renderer';

window.matchMedia = window.matchMedia || function () {
   return {
      matches: false,
      addListener: function () { },
      removeListener: function () { }
   };
};

describe('Comments', () => {

   it('Should successfully render', () => {
      const wrapper = renderer
         .create(
            <Comments />
         )
         .toJSON();
      expect(wrapper).toMatchSnapshot();
   });
});