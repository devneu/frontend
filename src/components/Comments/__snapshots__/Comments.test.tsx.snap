// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments Should successfully render 1`] = `
<div
  className="comments"
>
  <form
    className="ant-form ant-form-horizontal comments__form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <textarea
              className="ant-input"
              onChange={[Function]}
              onCompositionEnd={[Function]}
              onCompositionStart={[Function]}
              onKeyDown={[Function]}
              rows={4}
              style={Object {}}
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <button
              className="ant-btn ant-btn-primary"
              onClick={[Function]}
              type="submit"
            >
              <span>
                Add Comment
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
  <div
    className="ant-list ant-list-split comments__list"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-comment"
        >
          <div
            className="ant-comment-inner"
          >
            <div
              className="ant-comment-avatar"
            >
              <span
                className="ant-avatar ant-avatar-circle ant-avatar-image comments__commentor-avatar"
                style={Object {}}
              >
                <img
                  alt="Commentor Name"
                  onError={[Function]}
                  src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
                />
              </span>
            </div>
            <div
              className="ant-comment-content"
            >
              <div
                className="ant-comment-content-author"
              >
                <span
                  className="ant-comment-content-author-name"
                >
                  <span
                    className="comments__commentor-name"
                  >
                    Commenter Name
                  </span>
                </span>
              </div>
              <div
                className="ant-comment-content-detail"
              >
                <p>
                  We supply a series of design principles, practical patterns and high quality design resources (Sketch and Axure).
                </p>
              </div>
              <ul
                className="ant-comment-actions"
              >
                <li>
                  <span>
                    Reply to Commenter
                  </span>
                </li>
              </ul>
            </div>
          </div>
          <div
            className="ant-comment-nested"
          >
            <div
              className="ant-comment"
            >
              <div
                className="ant-comment-inner"
              >
                <div
                  className="ant-comment-avatar"
                >
                  <span
                    className="ant-avatar ant-avatar-circle ant-avatar-image comments__commentor-avatar"
                    style={Object {}}
                  >
                    <img
                      alt="Commentor Name"
                      onError={[Function]}
                      src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
                    />
                  </span>
                </div>
                <div
                  className="ant-comment-content"
                >
                  <div
                    className="ant-comment-content-author"
                  >
                    <span
                      className="ant-comment-content-author-name"
                    >
                      <span
                        className="comments__commentor-name"
                      >
                        Commenter Name
                      </span>
                    </span>
                  </div>
                  <div
                    className="ant-comment-content-detail"
                  >
                    <p>
                      We supply a series of design principles, practical patterns and high quality design resources (Sketch and Axure).
                    </p>
                  </div>
                  <ul
                    className="ant-comment-actions"
                  >
                    <li>
                      <span>
                        Reply to Commenter
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
            <div
              className="ant-comment"
            >
              <div
                className="ant-comment-inner"
              >
                <div
                  className="ant-comment-avatar"
                >
                  <span
                    className="ant-avatar ant-avatar-circle ant-avatar-image comments__commentor-avatar"
                    style={Object {}}
                  >
                    <img
                      alt="Commentor Name"
                      onError={[Function]}
                      src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
                    />
                  </span>
                </div>
                <div
                  className="ant-comment-content"
                >
                  <div
                    className="ant-comment-content-author"
                  >
                    <span
                      className="ant-comment-content-author-name"
                    >
                      <span
                        className="comments__commentor-name"
                      >
                        Commenter Name
                      </span>
                    </span>
                  </div>
                  <div
                    className="ant-comment-content-detail"
                  >
                    <p>
                      We supply a series of design principles, practical patterns and high quality design resources (Sketch and Axure).
                    </p>
                  </div>
                  <ul
                    className="ant-comment-actions"
                  >
                    <li>
                      <span>
                        Reply to Commenter
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="ant-comment"
        >
          <div
            className="ant-comment-inner"
          >
            <div
              className="ant-comment-avatar"
            >
              <span
                className="ant-avatar ant-avatar-circle ant-avatar-image comments__commentor-avatar"
                style={Object {}}
              >
                <img
                  alt="Commentor Name"
                  onError={[Function]}
                  src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
                />
              </span>
            </div>
            <div
              className="ant-comment-content"
            >
              <div
                className="ant-comment-content-author"
              >
                <span
                  className="ant-comment-content-author-name"
                >
                  <span
                    className="comments__commentor-name"
                  >
                    Commenter Name
                  </span>
                </span>
              </div>
              <div
                className="ant-comment-content-detail"
              >
                <p>
                  We supply a series of design principles, practical patterns and high quality design resources (Sketch and Axure).
                </p>
              </div>
              <ul
                className="ant-comment-actions"
              >
                <li>
                  <span>
                    Reply to Commenter
                  </span>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
